clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -447,10 +552,36 @@ public ConfigDef withClientSaslSupport() {
             String joined = Utils.join(undefinedConfigKeys, ",");
             throw new ConfigException("Some configurations in are referred in the dependents, but not defined: " + joined);
         }
-        // parse all known keys
+        // Handle all known ConfigKey objects.
         Map<String, Object> values = new HashMap<>();
-        for (ConfigKey key : configKeys.values())
-            values.put(key.name, parseValue(key, props.get(key.name), props.containsKey(key.name)));
+        for (ConfigKey key : configKeys.values()) {
+            if (values.containsKey(key.name))
+                continue;
+            Object value = null;
+            if (props.containsKey(key.name)) {
+                value = parseType(key.name, props.get(key.name), key.type);
+            } else {
+                boolean foundDeprecated = false;
+                for (String name : key.deprecatedNames) {
+                    if (props.containsKey(name)) {
+                        value = parseType(name, props.get(name), key.type);
+                        foundDeprecated = true;
+                        log.info("Configuration key {} is deprecated, and may be removed in the future.  " +
+                                "Please use {} instead.", name, key.name);
+                        break;
+                    }
+                }
+                if (!foundDeprecated) {
+                    if (key.defaultValue == NO_DEFAULT_VALUE)
```

Comment:
Nit: in `parseValue`, we changed this to `NO_DEFAULT_VALUE.equals(key.defaultValue)` due to findBugs warnings.